kmcd.md

```diff
@@ -80,6 +80,165 @@ State Storage/Revert Semantics
+ // -----------------------------------------------------------
+    rule unAuthFlipKick(
+           ListItem(FlipKick(ADDR, ILK, _, _, _, _, _, _))
+           EVENTS:List
+         )
+         => #if isAuthorized(ADDR, Flip ILK) #then unAuthFlipKick(EVENTS) #else true #fi
+
+    rule unAuthFlipKick( ListItem(_) EVENTS:List )
+         => unAuthFlipKick(EVENTS) [owise]
+
+    rule unAuthFlipKick(.List) => false
+```
+
+### Kicking off a fake `flap` auction (inspired by lucash-flap)
+
+The property checks if `flap . kick` is ever called by an unauthorized user (alternatively, the property can check whether a `flap` auction is kicked off with a zero bid?).
+
+```k
+    syntax Bool ::= unAuthFlapKick(List) [function, functional]
+ // -----------------------------------------------------------
+    rule unAuthFlapKick(
+          ListItem(FlapKick(ADDR, _, _, _))
+           EVENTS:List
+         )
+         => #if isAuthorized(ADDR, Flap) #then unAuthFlapKick(EVENTS) #else true #fi
+
+    rule unAuthFlapKick( ListItem(_) EVENTS:List )
+         => unAuthFlapKick(EVENTS) [owise]
+
+    rule unAuthFlapKick(.List) => false
+```
+
+### Earning interest from a pot after End is deactivated (inspired by the lucash-pot-end attack)
+
+The property checks if an `End . cage` is eventually followed by a successful `Pot . join`and `Pot . exit`.
+
+```k
+    syntax Bool ::= potEndInterest(List) [function, functional]
+ // -----------------------------------------------------------
+    rule potEndInterest(
+           ListItem(LogNote( ADDR, End . cage))
+           EVENTS:List
+         )
+         => potEndInterestBegin(EVENTS)
+
+    rule potEndInterest(ListItem(_) EVENTS:List )
+         => potEndInterest(EVENTS) [owise]
+
+    rule potEndInterest(.List) => false
+
+    syntax Bool ::= potEndInterestBegin(List) [function, functional]
+ // ----------------------------------------------------------------
+    rule potEndInterestBegin(
+           ListItem(LogNote( ADDR, Pot . join WAD ))
+           EVENTS:List
+         )
+         => potEndInterestEnd(EVENTS, ADDR)
+
+    rule potEndInterestBegin( ListItem(_) EVENTS:List )
+         => potEndInterestBegin(EVENTS) [owise]
+
+    rule potEndInterestBegin(.List) => false
+
+    syntax Bool ::= potEndInterestEnd(List, String) [function, functional]
+ // ----------------------------------------------------------------------
+    rule potEndInterestEnd(
+           ListItem(LogNote( ADDR, Pot . exit WAD ))
+           EVENTS:List, ADDR
+         )
+         => true
+
+    rule potEndInterestEnd( ListItem(_) EVENTS:List, ADDR )
+         => potEndInterestEnd(EVENTS, ADDR) [owise]
+
+    rule potEndInterestEnd(.List, _) => false
+```
+
+### Earning interest from a pot in zero time (inspired by the lucash-pot attack)
+
+The property checks if a sequence of `Pot . join`, `Pot . drip` and `Pot . exit` is executed in zero time.
+
+```k
+    syntax Bool ::= zeroTimePotInterest(List) [function, functional]
```

Comment:
A call to `Pot . join` which is preceded by a `TimeStep` more recently than a `Pot . drip` should not succeed.

1.  Check for calls to Pot . drip, if so go to 1. Check for calls to TimeStep, if so go to 2.
2.  Check for calls to Pot . join, if so violation. Check for calls to Pot . drip, if so go to 1.

Reply:
Done